### ISSUE TITLE
deep learning in six hours

### ISSUE THREAD

**Issue by changtimwu** [changtimwu/changtimwu.github.com#57]
http://www.slideshare.net/tw_dsconf/ss-62245351

**Comment by changtimwu** [comment]
p82 
original gradient descent vs mini-batch
- pro: mini-batch is much faster.  All batches can be parallely computed.
- con: mini-batch not toward minimum global loss

When you have a powerful parallel machine, do you still need mini-batch?
No!  No batch doesnt offer you better result eve in GTX 1080.

**Comment by changtimwu** [comment]
Remember to shuffle the batches!!

**Comment by changtimwu** [comment]
`keras` is quite handy.  just 3 lines of code to add 10-layers.

more layers doesn't guarantee better results.

vanishing gradient problem 梯度消失
靠近 input: smaller gradients
靠近 output: larger gradients
input 收斂太慢  
output 收斂很快

gradient 容易理解方式, 如果把參數 w 加個 delta, 對 loss function 變化有多到

現在都是用 ReLU 而不是 sigmod

**Comment by changtimwu** [comment]
discussion of ReLU
1. fast to compute
2. biologic reason 生物基礎
3. infinite sigmod with different basies
4. good against gradient vanish problem

把 output 為 0 的neron 移除, 剩下的都是 input=output
- becomes a thinner linear network
- do not have smaller gradients

本來 sigmod 會算到爛掉的, 換成 ReLU 可以得到 0.9571 accuracy.

**Comment by changtimwu** [comment]
MAXOUT: learnable activation fucntion
MAXOUT can approximate best activation function.  In such case, you do not bother the choice of activation function.

很特殊的 activation function 都可以被 maxout 調出來.
bias in MAXOUT?

Mr. Lee's personal exp.  Usually, you don't need normalization on ReLU.

**Comment by changtimwu** [comment]
Learning rates is a ke
1.  too large-> 跨過 min loss 沒法逼近
2. too small-> slow

不同參數不同  learning rate
parameter dependent learning rate

optimization: 
gradient descent:
local optimum

**Comment by changtimwu** [comment]
w range 太大
af input 常常都是極值, 

縱然 relu
接近轉折點
在 init 時候不要輸入太極端的值
盡量輸入在"轉折點"附近得直

**Comment by changtimwu** [comment]
how to overcome overfitting
- have more training data
- create more training data
- early stopping

weight decay
- if the link not used for a while, just adjust

**Comment by changtimwu** [comment]
dropout:
每次更新參數
every time updating parameter, drop some neurons and links it connected with
if dropout rate is p%, all the weights time (1-p)%

dropout 很神奇...
- dropout is kind of ensemble

dropout + maxout

dropout 加在 activation data

noisy training data + dropout

adding dropout  would not improve training data accuracy, it would only improve testing data set.

**Comment by changtimwu** [comment]
less parameters would be less chance over-fitting.

CNN:
靠近 input filter少 靠近output filter多

audio spectrum -> CNN

melody identification

**Comment by changtimwu** [comment]
feature engineering vs neural network structure engineering

deeping learning not good at text processing?

**Comment by changtimwu** [comment]
RNN training might be unstable.  Total loss may be volatile  through epochs.

**Comment by changtimwu** [comment]
entrolled to Hinton's class
https://www.coursera.org/learn/neural-networks/supplement/gKXYz/lecture-slides-and-resources

**Comment by changtimwu** [comment]
DL school 2016
https://www.youtube.com/watch?v=eyovmAtoUx0

**Comment by changtimwu** [comment]
just some typings to enhance memory.
Use CIFAR-10 as an example, each image is 32x32.  Then the input volume is [32x32x3], if the receptive field( or the filter size) is 5x5, then each neuron in the Conv Layer will have weights to
[5x5x3]

_For example for a 7x7 input and a 3x3 filter with stride 1 and pad 0 we would get a 5x5 output._
zero padding -> 不擴邊, 一次跳1
```
  1234567 
  123
   123
    123
     123
      123
```
(W−F+2P)/S +1= (7-3+2*0)/1 +1 = 4 + 1= 5
 

With stride 2 we would get a 3x3 output. L
不擴邊, 一次跳2
```
  1234567
  123
    123
      123
```
(W−F+2P)/S +1= (7-3+2*0)/2 +1 = 2 + 1= 3


http://www.songho.ca/dsp/convolution/convolution2d_example.html
最簡單的範例, 運算過程
3x3 input with filter 3x3, pad 1, stride 1
(W−F+2P)/S +1= (3-3+2*1)/1 +1 = 2/1 + 1= 3

**Comment by changtimwu** [comment]
let's see  how a filter work.  The filter 1 and filter 2 are effect vertical line  and horizontal line detectors.  That's why we need many filters to detect complex features.    Filters stack.  
https://www.youtube.com/watch?v=FrKWiRv254g&t=11m

6x6 input with filter 3x3, pad 0, stride 1
(W−F+2P)/S +1= (6-3+0*1)/1 +1 = 3/1 + 1= 4

You can view CNN as   fully connected layer cut with many connections.

 CNN 解釋性
please see t=43m,   找出某個 image 可以active 某個 filter 最大值 就可以理解描述這個 filter 的作用
filter 很多, 先抓前面12個, 某種紋路texure 的 detector

**Comment by changtimwu** [comment]
conv layer 的厚度是網路設計者訂出來的, 就是刻意要那麼多 layer, 通常會簡寫 ex. conv64 (3x3).
就代表 64 個 3x3 filter, 每個filter 對 input image 做 convolution 運算(掃一輪), 就會得到一個 64 層的 output.  
https://www.youtube.com/watch?v=FrKWiRv254g&t=29m
以 keras 為例
``` python
Convolution2D( 25, (3,3) , input_shape(1,28,28))
```
input 28x28 mono image, 25層的3x3 filter 
output shape keras 會幫你算
 
(W−F+2P)/S +1= (28-3+0*1)/1 +1 = 25/1 + 1= 26

就可以寫成
```
input: 1x28x28x1
conv: 25x3x3 
output: 25x26x26
```

下面有一個 conv
(W−F+2P)/S +1= (13-3+0*1)/1 +1 = 10/1 + 1= 11
```
input: 25x13x13
conv: 50x3x3
output: 50x11x11
```

25怎麼變到50??

**Comment by changtimwu** [comment]
```
INPUT: [224x224x3]        memory:  224*224*3=150K   weights: 0
CONV3-64: [224x224x64]  memory:  224*224*64=3.2M   weights: (3*3*3)*64 = 1,728
CONV3-64: [224x224x64]  memory:  224*224*64=3.2M   weights: (3*3*64)*64 = 36,864
POOL2: [112x112x64]  memory:  112*112*64=800K   weights: 0
CONV3-128: [112x112x128]  memory:  112*112*128=1.6M   weights: (3*3*64)*128 = 73,728
CONV3-128: [112x112x128]  memory:  112*112*128=1.6M   weights: (3*3*128)*128 = 147,456
POOL2: [56x56x128]  memory:  56*56*128=400K   weights: 0
CONV3-256: [56x56x256]  memory:  56*56*256=800K   weights: (3*3*128)*256 = 294,912
CONV3-256: [56x56x256]  memory:  56*56*256=800K   weights: (3*3*256)*256 = 589,824
CONV3-256: [56x56x256]  memory:  56*56*256=800K   weights: (3*3*256)*256 = 589,824
POOL2: [28x28x256]  memory:  28*28*256=200K   weights: 0
CONV3-512: [28x28x512]  memory:  28*28*512=400K   weights: (3*3*256)*512 = 1,179,648
CONV3-512: [28x28x512]  memory:  28*28*512=400K   weights: (3*3*512)*512 = 2,359,296
CONV3-512: [28x28x512]  memory:  28*28*512=400K   weights: (3*3*512)*512 = 2,359,296
POOL2: [14x14x512]  memory:  14*14*512=100K   weights: 0
CONV3-512: [14x14x512]  memory:  14*14*512=100K   weights: (3*3*512)*512 = 2,359,296
CONV3-512: [14x14x512]  memory:  14*14*512=100K   weights: (3*3*512)*512 = 2,359,296
CONV3-512: [14x14x512]  memory:  14*14*512=100K   weights: (3*3*512)*512 = 2,359,296
POOL2: [7x7x512]  memory:  7*7*512=25K  weights: 0
FC: [1x1x4096]  memory:  4096  weights: 7*7*512*4096 = 102,760,448
FC: [1x1x4096]  memory:  4096  weights: 4096*4096 = 16,777,216
FC: [1x1x1000]  memory:  1000 weights: 4096*1000 = 4,096,000
```

計算 number of parameters
把 weights 相加 = 138M

keras has `model.summary()`
```
Layer (type)                 Output Shape              Param #                                      
=================================================================                                   
zero_padding2d_1 (ZeroPaddin (None, 226, 226, 3)       0                                            
conv2d_1 (Conv2D)            (None, 224, 224, 64)      1792                                         
zero_padding2d_2 (ZeroPaddin (None, 226, 226, 64)      0                                            
conv2d_2 (Conv2D)            (None, 224, 224, 64)      36928                                        
max_pooling2d_1 (MaxPooling2 (None, 112, 112, 64)      0                                            
zero_padding2d_3 (ZeroPaddin (None, 114, 114, 64)      0                                            
conv2d_3 (Conv2D)            (None, 112, 112, 128)     73856                                        
zero_padding2d_4 (ZeroPaddin (None, 114, 114, 128)     0                                            
conv2d_4 (Conv2D)            (None, 112, 112, 128)     147584                                       
max_pooling2d_2 (MaxPooling2 (None, 56, 56, 128)       0                                            
zero_padding2d_5 (ZeroPaddin (None, 58, 58, 128)       0                                            
conv2d_5 (Conv2D)            (None, 56, 56, 256)       295168                                       
zero_padding2d_6 (ZeroPaddin (None, 58, 58, 256)       0  
zero_padding2d_7 (ZeroPaddin (None, 58, 58, 256)       0                                            
conv2d_7 (Conv2D)            (None, 56, 56, 256)       590080                                       
max_pooling2d_3 (MaxPooling2 (None, 28, 28, 256)       0                                            
zero_padding2d_8 (ZeroPaddin (None, 30, 30, 256)       0                                            
conv2d_8 (Conv2D)            (None, 28, 28, 512)       1180160                                      
zero_padding2d_9 (ZeroPaddin (None, 30, 30, 512)       0                                            
conv2d_9 (Conv2D)            (None, 28, 28, 512)       2359808                                      
zero_padding2d_10 (ZeroPaddi (None, 30, 30, 512)       0                                            
conv2d_10 (Conv2D)           (None, 28, 28, 512)       2359808                                      
max_pooling2d_4 (MaxPooling2 (None, 14, 14, 512)       0                                            
zero_padding2d_11 (ZeroPaddi (None, 16, 16, 512)       0                                            
conv2d_11 (Conv2D)           (None, 14, 14, 512)       2359808 
zero_padding2d_12 (ZeroPaddi (None, 16, 16, 512)       0                                            
conv2d_12 (Conv2D)           (None, 14, 14, 512)       2359808                                      
zero_padding2d_13 (ZeroPaddi (None, 16, 16, 512)       0                                            
conv2d_13 (Conv2D)           (None, 14, 14, 512)       2359808                                      
max_pooling2d_5 (MaxPooling2 (None, 7, 7, 512)         0                                            
flatten_1 (Flatten)          (None, 25088)             0                                            
dense_1 (Dense)              (None, 4096)              102764544                                    
dropout_1 (Dropout)          (None, 4096)              0                                            
dense_2 (Dense)              (None, 4096)              16781312                                     
dropout_2 (Dropout)          (None, 4096)              0                                            
dense_3 (Dense)              (None, 1000)              4097000                                      
=================================================================                                   
Total params: 138,357,544                                                                           
Trainable params: 138,357,544                                                                       
Non-trainable params: 0
```

**Comment by changtimwu** [comment]
computation cost of standard convolution
Dk*Dk*M*N*Df*Df

* Dk*Dk: kernel size
* M: number of input layers
* N: number of output layers
* Df*Df: output size

Take mobilenet's first layer(standard convolution, filter=3x3, stride=2)
* input shape: 224x224x3
* output shape: 112x112x32
==>
Dk=3, M=3, N=32, Df=112
